vmm/sandbox/src/cgroup.rs

```diff
@@ -0,0 +1,425 @@
+#[derive(Debug)]
+struct CpuSubsystem {
+    cpu_cgroup_root: PathBuf,
+}
+
+impl CpuSubsystem {
+    pub fn new() -> Self {
+        let path = cgroups_rs::hierarchies::V1::new()
+            .get_mount_point(cgroups_rs::Controllers::Cpu)
+            .unwrap();
+        CpuSubsystem {
+            cpu_cgroup_root: path,
+        }
+    }
+}
+
+pub fn cpu_subsystem_hier() -> Box<dyn Hierarchy> {
+    Box::new(CpuSubsystem::new())
+}
+
+impl Hierarchy for CpuSubsystem {
+    fn subsystems(&self) -> Vec<cgroups_rs::Subsystem> {
+        vec![Subsystem::Cpu(CpuController::new(
+            self.cpu_cgroup_root.clone(),
+            false,
+        ))]
+    }
+
+    fn root(&self) -> std::path::PathBuf {
+        self.cpu_cgroup_root.parent().unwrap().to_path_buf()
+    }
+
+    fn root_control_group(&self) -> Cgroup {
+        Cgroup::load(cpu_subsystem_hier(), "")
+    }
+
+    fn parent_control_group(&self, path: &str) -> Cgroup {
+        let path = Path::new(path);
+        let parent_path = path.parent().unwrap().to_string_lossy().to_string();
+        Cgroup::load(cpu_subsystem_hier(), parent_path)
+    }
+
+    fn v2(&self) -> bool {
+        false
+    }
+}
+
+fn apply_cpu_resource(cgroup: &Cgroup, res: &LinuxContainerResources) -> Result<()> {
+    let cpu_controller: &CpuController = cgroup
+        .controller_of()
+        .ok_or(anyhow!("No cpu controller attached!"))?;
+    cpu_controller.set_cfs_period(res.cpu_period.try_into()?)?;
+    cpu_controller.set_cfs_quota(res.cpu_quota)?;
+    cpu_controller.set_shares(res.cpu_shares.try_into()?)?;
+    Ok(())
+}
+
+fn apply_memory_resource(cgroup: &Cgroup, res: &LinuxContainerResources) -> Result<()> {
+    let mem_controller: &MemController = cgroup
+        .controller_of()
+        .ok_or(anyhow!("No memory controller attached!"))?;
+    mem_controller.set_limit(res.memory_limit_in_bytes)?;
+    mem_controller.set_memswap_limit(res.memory_swap_limit_in_bytes)?;
+    Ok(())
+}
+
+fn apply_cpuset_resources(cgroup: &Cgroup, res: &LinuxContainerResources) -> Result<()> {
+    let cpuset_controller: &CpuSetController = cgroup
+        .controller_of()
+        .ok_or(anyhow!("No cpuset controller attached!"))?;
+    cpuset_controller.set_cpus(&res.cpuset_cpus)?;
+    cpuset_controller.set_mems(&res.cpuset_mems)?;
+    Ok(())
+}
+
+fn apply_hugetlb_resources(cgroup: &Cgroup, res: &LinuxContainerResources) -> Result<()> {
+    let hugetlb_controller: &HugeTlbController = cgroup
+        .controller_of()
+        .ok_or(anyhow!("No hugetlb controller attached!"))?;
+    for h in res.hugepage_limits.iter() {
+        hugetlb_controller.set_limit_in_bytes(h.page_size.as_str(), h.limit)?;
+    }
+    Ok(())
+}
+
+fn remove_sandbox_cgroup(cgroup: &Cgroup) -> Result<()> {
+    // get the tids in the current cgroup and then move the tids to parent cgroup
+    let tids = cgroup.tasks();
+    for tid in tids {
+        cgroup.move_task_to_parent(tid)?;
```

Comment:
Should these tasks be moved out or killed? cc @abel-von

Reply:
The default operation before removing a cgroup dir is to move the processes to its parent directory, because some processes may not be killed due to the D state.